selfservice/hook/web_hook.go

```diff
@@ -0,0 +1,307 @@
+package hook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/ory/kratos/identity"
+	"github.com/ory/kratos/selfservice/flow/settings"
+
+	"github.com/ory/kratos/selfservice/flow/login"
+
+	"github.com/ory/kratos/selfservice/flow/recovery"
+
+	"github.com/google/go-jsonnet"
+
+	"github.com/ory/kratos/selfservice/flow/registration"
+	"github.com/ory/kratos/selfservice/flow/verification"
+	"github.com/ory/kratos/session"
+	"github.com/ory/kratos/x"
+)
+
+var _ registration.PostHookPostPersistExecutor = new(WebHook)
+var _ verification.PostHookExecutor = new(WebHook)
+var _ recovery.PostHookExecutor = new(WebHook)
+
+type (
+	AuthStrategy interface {
+		apply(req *http.Request)
+	}
+
+	authStrategyFactory func(c json.RawMessage) (AuthStrategy, error)
+
+	noopAuthStrategy struct{}
+
+	basicAuthStrategy struct {
+		user     string
+		password string
+	}
+
+	apiKeyStrategy struct {
+		name  string
+		value string
+		in    string
+	}
+
+	webHookConfig struct {
+		method       string
+		url          string
+		templatePath string
+		auth         AuthStrategy
+	}
+
+	webHookDependencies interface {
+		x.LoggingProvider
+	}
+
+	templateContext struct {
+		Flow           interface{} `json:"flow"`
```

Comment:
I think it would make sense to add the flow name as well? So `login`, `registration`, and so on. That helps us differentiate in the webhook jsonnet!

Reply:
Agree. IMHO, it would be better to have the corresponding property in the `Flow` object itself. One would then get the name of the Flow in the ctx object just like this: `ctx.flow.name`. This would however require an update of at least the open api spec. I mean there is no need to write that information to the db, as this information is static. There are however multiple places, which fetch the flow data from the db and would require an update. It would be great, if you could support here. I can then focus entirely on the tests for the web hooks.

Reply:
That makes sense! I think it would probably be easier to just do the type assertion in the JsonNet config here - but we can do that in a later PR.